File: SoftwareArchitecture/Principle/SOLID/04-InterfaceSegregationPrinciple/Animal/solution/Animal.java
public interface Flyable {
    void fly();
}
public interface Runnable {
    void run();
}
public interface Barkable {
    void bark();
}
public class Bird implements Flyable, Runnable {
    public void run() {
        // write code about running of the bird
    }
    public void fly() {
        // write code about flying of the bird
    }
}
public class Cat implements Runnable{
    public void run() {
        // write code about running of the cat
    }
}
public class Dog implements Runnable, Barkable {
    public void bark() {
        // write code about barking of the dog
    }
    public void run() {
        // write code about running of the dog
    }
}

File: SoftwareArchitecture/Principle/SOLID/04-InterfaceSegregationPrinciple/AnimalGo/Animal.go
package main

import "fmt"

type Animal struct {
	ID   int
	Name string
}

type Flyable interface {
	fly()
}

type Runable interface {
	run()
}

type Barkable interface {
	bark()
}

type Cat struct {
	Animal
}

func (c Cat) run() {
	fmt.Println("Cat is running")
}

type Dog struct {
	Animal
}

func (d Dog) run() {
	fmt.Println("Cat is running")
}

func (d Dog) bark() {
	fmt.Println("Cat is barkin")
}

type Bird struct {
	Animal
}

func (b Bird) bark() {
	fmt.Println("Bird is leo lo leo lo")
}

func (b Bird) fly() {
	fmt.Println("Bird is flying")
}

func main() {
	cat := Cat{Animal: Animal{
		ID:   1,
		Name: "cat",
	}}
	cat.run()
	fmt.Println(cat)

}
